contracts/zero-ex/contracts/src/features/ERC721OrdersFeature.sol

```diff
@@ -0,0 +1,1112 @@
+                    spread
+                )
+            );
+            // Step 7: The spread less the sell order fees is the amount of ETH
+            //         remaining in the EP that can be sent to `msg.sender` as
+            //         the profit from matching these two orders.
+            profit = spread - sellOrderFees;
+            if (profit > 0) {
+                _transferEth(msg.sender, profit);
+            }
+        } else {
+            // Step 1: Transfer the ERC20 token from the buyer to the seller.
+            //         Note that we transfer `sellOrder.erc20TokenAmount`, which
+            //         is at most `buyOrder.erc20TokenAmount`.
+            _transferERC20TokensFrom(
+                buyOrder.erc20Token,
+                buyOrder.maker,
+                sellOrder.maker,
+                sellOrder.erc20TokenAmount
+            );
+
+            // Step 2: Pay fees for the buy order. Note that these are paid
+            //         by the buyer. By signing the buy order, the buyer signals
+            //         that they are willing to spend a total of 
+            //         `buyOrder.erc20TokenAmount` _plus_ `buyOrder.fees`.
+            _payFees(buyOrder, buyOrder.maker, false);
+
+            // Step 3: Pay fees for the sell order. These are paid by the buyer
+            //         as well. After paying these fees, we may have taken more
+            //         from the buyer than they agreed to in the buy order. If
+            //         so, we revert in the following step.
+            uint256 sellOrderFees = _payFees(sellOrder, buyOrder.maker, false);
+            
+            // Step 4: The spread must be enough to cover the sell order fees.
+            //         If not, `_payFees` will have taken more tokens from the 
+            //         buyer than they had agreed to in the buy order, in which 
+            //         case we revert here.
+            rrequire(
+                spread >= sellOrderFees,
+                LibERC721OrdersRichErrors.SellOrderFeesExceedSpreadError(
+                    sellOrderFees,
+                    spread
+                )
+            );
+
+            // Step 7: We calculate the profit as:
+            //         profit = buyOrder.erc20TokenAmount - sellOrder.erc20TokenAmount - sellOrderFees
+            //                = spread - sellOrderFees
+            //         I.e. the buyer would've been willing to pay up to `profit`
+            //         more to buy the asset, so instead that amount is sent to
+            //         `msg.sender` as the profit from matching these two orders.
+            profit = spread - sellOrderFees;
+            if (profit > 0) {
+                _transferERC20TokensFrom(
+                    buyOrder.erc20Token,
+                    buyOrder.maker,
+                    msg.sender,
+                    profit
+                );
+            }
+        }
+
+        // TODO: or we could emit 2 ERC721OrderFilled events
```

Comment:
I like that. Maybe those events could also have a `matcher` param to differentiate.